File: ingest/scoring/spothours.py
"""Each spot's own numbers, hour by hour, as a byte per sample.

Why a blob and not five arrays inside conditions.json: ten days is 240 hours
for each of fifteen spots, and written as JSON that is some sixty kilobytes
against a document that is currently twenty-eight. `test_build.py` puts a 35KB
ceiling on that document and the reason is in its docstring - it is fetched
over mobile data on a mountain road, which is exactly the moment this product
is for. Quantised to a byte a sample the same series is eighteen kilobytes, it
travels beside conditions.json rather than inside it, and it can fail to load
without taking the scores down with it, the same way the cloud volume and the
observed field already do.

A byte is not a lossy shortcut here, it is honest about the input. The pressure
levels near the summits are roughly five hundred metres apart and the deck
altitude is interpolated between them, so twenty-metre steps are already finer
than the data can support.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from typing import Sequence

# The byte that means "no value here". Every series gives it up, which is why
# none of them may use the full 0..255 range.
MISSING = 255


@dataclass(frozen=True)
class Series:
    """One quantised channel: `value = byte * scale + offset`."""

    name: str
    scale: float
    offset: float

    def encode(self, value: float | None) -> int:
        if value is None:
            return MISSING
        raw = round((value - self.offset) / self.scale)
        # Clamped rather than wrapped. A wind of 300km/h is wrong, but drawing
        # it as 45 because the byte overflowed is worse than drawing the
        # highest wind the scale can hold.
        return max(0, min(MISSING - 1, raw))

    def decode(self, raw: int) -> float | None:
        return None if raw == MISSING else raw * self.scale + self.offset


# Ranges chosen from what the archipelago actually produces, with room above.
SERIES: tuple[Series, ...] = (
    # 0..5080m in 20m steps: the ladder tops out at 3000m and the levels near
    # the summits are 500m apart.
    Series("deck_base_m", 20.0, 0.0),
    Series("deck_top_m", 20.0, 0.0),
    # 0..1 in half-percent steps.
    Series("cloud_at_summit", 0.005, 0.0),
    # -25C to +38C in quarter degrees. Madeira's summits go below zero in
    # winter and the coast does not reach thirty-nine.
    Series("temperature_c", 0.25, -25.0),
    # 0..127km/h in half-km steps.
    Series("wind_kmh", 0.5, 0.0),
    # Aerosol optical depth, 0..2.54 in hundredths. Clean air here is near
    # 0.10 and a heavy calima is past 1.0, so hundredths is finer than the
    # model's own disagreement about the number.
    Series("aod", 0.01, 0.0),
)


@dataclass(frozen=True)
class SpotHours:
    """The whole grid of it: [spot][series][hour], one byte each."""

    spot_ids: tuple[str, ...]
    t0: datetime
    count: int
    values: bytes

    @property
    def stride(self) -> int:
        """Bytes per spot."""
        return len(SERIES) * self.count


def encode(
    spot_ids: Sequence[str],
    t0: datetime,
    count: int,
    columns: dict[str, dict[str, Sequence[float | None]]],
) -> SpotHours:
    """Pack `columns[spot_id][series_name]` into one blob.

    A spot with no forecast is not dropped from the layout - it is written as a
    run of MISSING - because the index a reader computes from a spot's position
    has to stay valid whether or not that particular fetch succeeded.
    """
    payload = bytearray()
    for spot_id in spot_ids:
        spot = columns.get(spot_id, {})
        for series in SERIES:
            samples = spot.get(series.name, ())
            if samples and len(samples) != count:
                raise ValueError(
                    f"{spot_id}/{series.name}: {len(samples)} samples for {count} hours"
                )
            if not samples:
                payload += bytes([MISSING]) * count
                continue
            payload += bytes(series.encode(value) for value in samples)

    return SpotHours(
        spot_ids=tuple(spot_ids),
        t0=t0,
        count=count,
        values=bytes(payload),
    )


def header(hours: SpotHours, generated_at: datetime) -> dict:
    """What travels inside conditions.json to describe the blob beside it."""
    return {
        "file": "spot-hours.bin",
        "generated_at": generated_at.astimezone(timezone.utc)
        .isoformat()
        .replace("+00:00", "Z"),
        "t0": hours.t0.astimezone(timezone.utc).isoformat().replace("+00:00", "Z"),
        "step_h": 1,
        "count": hours.count,
        # Order is the index. A reader finds a spot's block by its position
        # here, so this list and the blob are one artefact.
        "spots": list(hours.spot_ids),
        "series": [
            {"name": series.name, "scale": series.scale, "offset": series.offset}
            for series in SERIES
        ],
        "missing": MISSING,
        "bytes": len(hours.values),
    }


def value_at(
    hours: SpotHours, spot_id: str, series_name: str, when: datetime
) -> float | None:
    """One number, for the tests and the dry run. The web side has its own."""
    if spot_id not in hours.spot_ids:
        return None
    series_index = next(
        (index for index, s in enumerate(SERIES) if s.name == series_name), None
    )
    if series_index is None:
        return None

    elapsed = when.astimezone(timezone.utc) - hours.t0
    index = round(elapsed / timedelta(hours=1))
    if index < 0 or index >= hours.count:
        return None

    spot_index = hours.spot_ids.index(spot_id)
    at = spot_index * hours.stride + series_index * hours.count + index
    return SERIES[series_index].decode(hours.values[at])

File: web/lib/spotHours.ts
/**
 * Each spot's own numbers, hour by hour.
 *
 * Without this the scrubber moved the map and nothing else: the volume, the
 * heatmap, the observed field and the sun all followed the hour, while the
 * sidebar, the sheet and the callout kept showing the *day* summary. Scrub to
 * 04:00 and the picture and the panel beside it were answering different
 * questions - which is the one failure this product cannot afford, because the
 * whole pitch is that the number and the picture agree.
 *
 * Shipped as a byte per sample beside conditions.json, for the reason stated
 * in ingest/scoring/spothours.py: as JSON the same series is sixty kilobytes
 * against a document that is meant to stay under thirty-five, and it is
 * fetched over mobile data on a mountain road.
 */

export type SpotHoursSeries = {
  name: string;
  /** `value = byte * scale + offset`. */
  scale: number;
  offset: number;
};

export type SpotHoursHeader = {
  file: string;
  generated_at: string;
  /** First hour, UTC. The series is contiguous from here at `step_h`. */
  t0: string;
  step_h: number;
  count: number;
  /** Order is the index: a spot's block is found by its position here. */
  spots: string[];
  series: SpotHoursSeries[];
  /** The byte that means "no value", to be read as null rather than as zero. */
  missing: number;
  bytes: number;
};

export type SpotHours = {
  header: SpotHoursHeader;
  values: Uint8Array;
  t0Ms: number;
};

/** What one hour says about one spot. Any field may be missing. */
export type SpotHour = {
  deckBaseM: number | null;
  deckTopM: number | null;
  cloudAtSummit: number | null;
  temperatureC: number | null;
  windKmh: number | null;
  /** Aerosol optical depth: Saharan dust, which hazes a view the vertical
   *  profile cannot see at all. */
  aod: number | null;
};

export function expectedBytes(header: SpotHoursHeader): number {
  return header.spots.length * header.series.length * header.count;
}

export function spotHoursUrl(
  header: SpotHoursHeader,
  documentUrl: string,
): string {
  const base = documentUrl.slice(0, documentUrl.lastIndexOf("/") + 1);
  return `${base}${header.file}`;
}

export function decodeSpotHours(
  header: SpotHoursHeader,
  values: Uint8Array,
): SpotHours {
  const expected = expectedBytes(header);
  if (header.bytes !== expected) {
    throw new Error(
      `spot hours header declares ${header.bytes} bytes for ${expected} samples`,
    );
  }
  if (values.length !== expected) {
    throw new Error(
      `spot hours is ${values.length} bytes, expected ${expected}`,
    );
  }
  return { header, values, t0Ms: new Date(header.t0).getTime() };
}

export async function loadSpotHours(
  header: SpotHoursHeader,
  documentUrl: string,
  fetchImpl: typeof fetch = fetch,
): Promise<SpotHours> {
  const response = await fetchImpl(spotHoursUrl(header, documentUrl), {
    cache: "no-store",
  });
  if (!response.ok) throw new Error(`spot hours: HTTP ${response.status}`);
  const buffer = await response.arrayBuffer();
  return decodeSpotHours(header, new Uint8Array(buffer));
}

/**
 * The index of the hour covering an instant, or null when it is outside what
 * was published.
 *
 * Null rather than the nearest edge. Both the forecast and the archive end,
 * and holding the last hour in place while the scrubber runs past it would
 * present one hour's weather as another's.
 */
export function hourIndexAt(hours: SpotHours, atMs: number): number | null {
  const step = hours.header.step_h * 3600_000;
  const index = Math.round((atMs - hours.t0Ms) / step);
  return index < 0 || index >= hours.header.count ? null : index;
}

function seriesValue(
  hours: SpotHours,
  spotIndex: number,
  seriesIndex: number,
  hourIndex: number,
): number | null {
  const { series, count, missing } = hours.header;
  const at =
    spotIndex * series.length * count + seriesIndex * count + hourIndex;
  const raw = hours.values[at];
  if (raw === missing) return null;
  return raw * series[seriesIndex].scale + series[seriesIndex].offset;
}

/**
 * Everything published for one spot at one instant, or null when that spot or
 * that hour is not in the blob.
 *
 * The series are looked up by name rather than by position, so the ingest can
 * add one without this reading the wrong channel until both sides are
 * redeployed.
 */
export function spotHourAt(
  hours: SpotHours,
  spotId: string,
  atMs: number,
): SpotHour | null {
  const spotIndex = hours.header.spots.indexOf(spotId);
  if (spotIndex < 0) return null;
  const hourIndex = hourIndexAt(hours, atMs);
  if (hourIndex === null) return null;

  const read = (name: string) => {
    const seriesIndex = hours.header.series.findIndex(
      (series) => series.name === name,
    );
    if (seriesIndex < 0) return null;
    return seriesValue(hours, spotIndex, seriesIndex, hourIndex);
  };

  return {
    deckBaseM: read("deck_base_m"),
    deckTopM: read("deck_top_m"),
    cloudAtSummit: read("cloud_at_summit"),
    temperatureC: read("temperature_c"),
    windKmh: read("wind_kmh"),
    aod: read("aod"),
  };
}

/** The span the blob covers, as epoch milliseconds. */
export function coverage(hours: SpotHours): { fromMs: number; toMs: number } {
  const step = hours.header.step_h * 3600_000;
  return {
    fromMs: hours.t0Ms,
    toMs: hours.t0Ms + (hours.header.count - 1) * step,
  };
}
